### PR TITLE
Add World serialization example

### PR DESCRIPTION
**Pull request by kaphula**
As discussed in https://github.com/Ralith/hecs/issues/380

### CODE REVIEW

File: examples/serialize_to_disk.rs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+//! a save mechanism for a game.
+//!
+//! The example creates a sample `World`, serializes it, saves it to disk as `saved_world.world` file,
+//! loads it back from the disk, deserializes the loaded world data and validates that component
+//! data of the worlds match.
+//!
+//! Run this example from crate root with:
+//! `cargo run --example serialize_to_disk --features "column-serialize"`
+
+#[cfg(feature = "column-serialize")]
```
Comment (Ralith) [inline examples/serialize_to_disk.rs:12]: I'm not sure what the point of this cfg gate is, because this will fail to compile regardless if the feature is disabled. Instead, define the feature as required for the example in Cargo.toml.

File: examples/serialize_to_disk.rs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+mod serialize_to_disk_example {
```
Comment (Ralith) [inline examples/serialize_to_disk.rs:13]: Why have an inner module here?

File: Cargo.toml [hunk 3]
```diff
@@ -41,6 +41,7 @@ rand = "0.8.3"
 trybuild = "1.0.23"
 serde = { version = "1.0.117", features = ["derive"] }
 serde_test = "1.0.117"
+serde_json = "1.0.117"
```
Comment (Ralith) [inline Cargo.toml:44]: JSON is a poor choice to illustrate column-major serialization. If you want the output to be human-readable, use row-major serialization. If you don't, don't use a slow, complex, text-based format.

File: .gitignore [hunk 4]
```diff
@@ -1,3 +1,4 @@
 /target
 **/*.rs.bk
 Cargo.lock
+.idea
```
Comment (Ralith) [inline .gitignore:4]: Unrelated change.

**Review by Ralith — CHANGES_REQUESTED** [review]

File: Cargo.toml (hunk 3 above)
Comment (kaphula) [reply Cargo.toml:44]: Okay, should I use bincode instead or something else?

File: examples/serialize_to_disk.rs (hunk 1 above)
Comment (kaphula) [reply examples/serialize_to_disk.rs:12]: True!

File: Cargo.toml (hunk 3 above)
Comment (Ralith) [reply Cargo.toml:44]: If you want to illustrate columnar serialization, bincode would be a reasonable choice.

**Comment by kaphula** [comment]
I applied the fixes you suggested. Does it look better now?

File: examples/serialize_to_disk.rs (hunk 1 above)
Comment (Ralith) [reply examples/serialize_to_disk.rs]: This comment has not been addressed.

File: examples/serialize_to_disk.rs [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+                ComponentId::ComponentB => {
+                    deserialize_column::<ComponentB, _>(entity_count, &mut seq, batch)?;
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl SerializeContext for SaveContext {
```
Comment (Ralith) [inline examples/serialize_to_disk.rs:93]: This shouldn't be implemented on the same type as `DeserializeContext`, because it doesn't need the same (or any) data on the side.

**Review by Ralith — CHANGES_REQUESTED** [review]

File: examples/serialize_to_disk.rs (hunk 5 above)
Comment (kaphula) [reply examples/serialize_to_disk.rs:93]: So should it be done like this?

```rust
#[derive(Default)]
pub struct SaveContextSerialize {}
#[derive(Default)]
pub struct SaveContextDeserialize {
   pub components: Vec<ComponentId>,
}
```

Comment (Ralith) [reply examples/serialize_to_disk.rs:93]: That would be clearer, yes.

File: examples/serialize_to_disk.rs [hunk 6]
```diff
@@ -45,7 +47,7 @@ pub struct ComponentB {
 }
 
 #[cfg(feature = "column-serialize")]
```
Comment (Ralith) [inline examples/serialize_to_disk.rs:49]: Still got a stray cfg.

File: examples/serialize_to_disk.rs (hunk 1 above)
Comment (kaphula) [reply examples/serialize_to_disk.rs]: Is it fine now?

File: examples/serialize_to_disk.rs (hunk 6 above)
Comment (kaphula) [reply examples/serialize_to_disk.rs]: Removed.

File: examples/serialize_to_disk.rs [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+            // the same:
+            println!("Loaded world \'{}\' from disk.", path.display());
+
+            print!("Validating world data... ");
+            for (_, t) in &mut world.query::<&ComponentA>() {
+                assert_eq!(t, &input_data1);
+            }
+            for (_, t) in &mut world.query::<&ComponentB>() {
+                assert_eq!(t, &input_data2);
+            }
```
Comment (Ralith) [inline examples/serialize_to_disk.rs:173]: You should also check `world.len()`, or this will pass if the world is empty.

**Review by Ralith — APPROVED** [review]

**Comment by Ralith** [comment]
Thanks!
